test/libsolidity/ASTPropertyTest.cpp

```diff
@@ -0,0 +1,202 @@
+		BOOST_THROW_EXCEPTION(runtime_error("Not a Solidity file: \"" + _filename + "\"."));
+
+	m_source = m_reader.source();
+	readExpectations();
+}
+
+void ASTPropertyTest::generateTestCaseValues(string& _values, bool _obtained)
+{
+	_values.clear();
+	for (string const& testId: m_expectationsSequence)
+	{
+		soltestAssert(m_testCases.count(testId) > 0);
+		_values +=
+			testId +
+			": " +
+			(_obtained ? m_testCases[testId].obtainedValue : m_testCases[testId].expectedValue)
+			+ "\n";
+	}
+}
+
+vector<StringPair> ASTPropertyTest::readKeyValuePairs(string const& _input)
+{
+	vector<StringPair> result;
+	for (string const& line: _input | ranges::views::split('\n') | ranges::to<vector<string>>)
+	{
+		if (line.empty())
+			continue;
+
+		vector<string> pair;
+		for (string const& element: line | ranges::views::split(':') | ranges::to<vector<string>>)
+		{
+			soltestAssert(ranges::all_of(boost::trim_copy(element), [](char c) { return isprint(c); }));
+			pair.emplace_back(boost::trim_copy(element));
+		}
+
+		soltestAssert(pair.size() == 2, "Invalid property test: " + line);
+		result.emplace_back(pair[0], pair[1]);
+	}
+	return result;
+}
+
+void ASTPropertyTest::readExpectations()
+{
+	for (auto const& [testId, testExpectation]: readKeyValuePairs(m_reader.simpleExpectations()))
+	{
+		soltestAssert(m_testCases.count(testId) == 0, "More than one expectation for test \"" + testId + "\"");
+		m_testCases.emplace(testId, ASTPropertyTestCase{testId, "", testExpectation, ""});
+		m_expectationsSequence.push_back(testId);
+	}
+	generateTestCaseValues(m_expectation, false);
+}
+
+optional<Json::Value> ASTPropertyTest::findNode(Json::Value const& _root, string_view const& _property)
+{
+	if (_root.isObject() && !_property.empty())
+	{
+		string subNode = string(_property.substr(0, _property.find_first_of('.')));
+		if (subNode != _property)
+			return findNode(_root[subNode], _property.substr(_property.find_first_of('.') + 1));
+		if (_root.isMember(subNode))
+			return make_optional(_root[subNode]);
+	}
+	return {};
+}
+
+void ASTPropertyTest::readTestedProperties(Json::Value const& _astJson)
+{
+	queue<Json::Value> nodesToVisit;
+	nodesToVisit.push(_astJson);
+
+	set<string> readTests;
+
+	while (!nodesToVisit.empty())
+	{
+		Json::Value& node = nodesToVisit.front();
+
+		if (node.isArray())
+			for (auto&& member: node)
+				nodesToVisit.push(member);
+		else if (node.isObject())
+			for (string const& memberName: node.getMemberNames())
+			{
+				if (memberName != "documentation")
+				{
+					nodesToVisit.push(node[memberName]);
+					continue;
+				}
+
+				string nodeDocstring = node["documentation"].isObject() ?
+					node["documentation"]["text"].asString() :
+					node["documentation"].asString();
+				soltestAssert(!nodeDocstring.empty());
+
+				vector<StringPair> pairs = readKeyValuePairs(nodeDocstring);
+				soltestAssert(!pairs.empty(), "It is required at least one property test per annotated AST node.");
```

Comment:
This assert will catch the case when there is an empty line (actually a line with spaces only) between two valid test lines:
```
/// Test1: name
///    
/// Test2: visibility
```
If the line is really empty, i.e, no spaces at all, then it doesn't show in the ast JSON.

Reply:
Hmm... why would that check catch it in the case you're showing? I'd expect `pairs.empty()` to be true because you'll still have `Test1: name` and `Test2: visibility` in the vector.

Reply:
Yeah, true. I guess I was thinking of something else when I wrote that... :smile: 
It will actually catch the case when all of the lines are blank.

Reply:
We could also just ignore and continue processing... but would it make sense to have a blank test?

Reply:
It would not be terribly useful, but I don't think it should be an error. I actually think it would be marginally useful to even have Natspec comments that are not tests, but that's a bigger change and not worth it in terms of effort. Ignoring empty ones is something we can very easily do though.

Reply:
Ok. It now ignores empty/blank tests, i.e, all lines are empty or have only whitespaces.

Reply:
Ah, I just realized what you were worried about. I see the test with no expectations at all :)

Well, you should be able to have empty Natspec lines, but it would still be useful to enforce that the file as a whole has at least one expectation defined. That's what we do for example in semantic tests. Could you add an assert at the end of the function, checking that there is at least one expectation line at the end?

Also, would be helpful to assert that all defined properties are actually used in expectations. If not, the person writing the case most likely forgot one of them.

Reply:
> That was already implemented. Any tests with no corresponding property or expectation produce an assertion fail.

I meant checking at there's at least one test present below `// ----`.

Reply:
> That was already implemented. Any tests with no corresponding property or expectation produce an assertion fail.

Ah, right. The `does not have a corresponding expected value` check does cover that.